### PR TITLE
feat(infra): use cache in-memory cluster for ws socket.io adapter

### PR DESCRIPTION
**Pull request by p-fernandez**
### What change does this PR introduce?

WS Socket.io adapter will use a new InMemoryServiceProvider so in the future we can isolate it to a dedicate InMemoryProvider.

### Why was this change needed?

We are decommissioning old Redis cluster instance.

### CODE REVIEW

File: apps/ws/src/shared/framework/in-memory-io.adapter.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { IoAdapter } from '@nestjs/platform-socket.io';
+import { ServerOptions } from 'socket.io';
+import Redis from 'ioredis';
+import { createAdapter } from '@socket.io/redis-adapter';
+import { getRedisPrefix } from '@novu/shared';
+import { WebSocketsInMemoryProviderService } from '@novu/application-generic';
+import { ConnectionOptions } from 'tls';
```
Comment (LetItRock) [inline apps/ws/src/shared/framework/in-memory-io.adapter.ts:7]: some imports are not used

File: apps/ws/src/shared/framework/in-memory-io.adapter.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+export class InMemoryIoAdapter extends IoAdapter {
+  private webSocketsInMemoryProviderService: WebSocketsInMemoryProviderService;
+  private adapterConstructor: ReturnType<typeof createAdapter>;
+
+  async connectToInMemoryCluster(): Promise<void> {
+    // TODO: Pending to inject in the provider instantiation
+    const keyPrefix = getRedisPrefix() ? `socket.io#${getRedisPrefix()}` : 'socket.io';
```
Comment (LetItRock) [inline apps/ws/src/shared/framework/in-memory-io.adapter.ts:15]: do you plan to make this change in a separate PR?

**Review by LetItRock — APPROVED** [review]
looks good to me 👍

File: apps/ws/src/shared/framework/in-memory-io.adapter.ts (hunk 2 above)
Comment (p-fernandez) [reply apps/ws/src/shared/framework/in-memory-io.adapter.ts:15]: This change is dependant and also this PR on this other PR: https://github.com/novuhq/novu/pull/4868
I gave the full context in the Linear task.

Comment (LetItRock) [reply apps/ws/src/shared/framework/in-memory-io.adapter.ts:15]: ok, thanks, will take a look into that PR ;)

Comment (LetItRock) [reply apps/ws/src/shared/framework/in-memory-io.adapter.ts:15]: maybe then rebase this one to `nv-3197-move-provider-selection-to-workflow-and-cache-in-memory`?

**Review by djabarovgeorge — COMMENTED** [review]
Looks good had just one question.

File: apps/ws/src/shared/framework/in-memory-io.adapter.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    this.webSocketsInMemoryProviderService = new WebSocketsInMemoryProviderService();
+
+    await this.webSocketsInMemoryProviderService.initialize();
+
+    const pubClient = this.webSocketsInMemoryProviderService.getClient();
+    const subClient = pubClient?.duplicate();
+
+    /*
+     *  TODO: Might not be needed to connect as we are checking it is initialized already.
+     * await Promise.all([pubClient?.connect(), subClient?.connect()]);
```
[the remark this answers is not in the document]
Comment (p-fernandez) [reply apps/ws/src/shared/framework/in-memory-io.adapter.ts:26]: This code was existing before my changes. If I leave it, now because the InMemoryProviderService downstream is initialised if you try to connect again it errors.

File: apps/ws/src/shared/framework/in-memory-io.adapter.ts (hunk 2 above)
Comment (p-fernandez) [reply apps/ws/src/shared/framework/in-memory-io.adapter.ts:15]: Yes, I will do that once NV-3197 is reviewed and approved.

File: packages/application-generic/src/services/in-memory-provider/web-sockets-in-memory-provider.service.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+export class WebSocketsInMemoryProviderService {
+  public inMemoryProviderService: InMemoryProviderService;
+  public isCluster: boolean;
+  private getIsInMemoryClusterModeEnabled: GetIsInMemoryClusterModeEnabled;
+
+  constructor() {
+    this.getIsInMemoryClusterModeEnabled =
+      new GetIsInMemoryClusterModeEnabled();
+
+    const provider = this.selectProvider();
```
[the remark this answers is not in the document]
Comment (p-fernandez) [reply packages/application-generic/src/services/in-memory-provider/web-sockets-in-memory-provider.service.ts:23]: The log will disappear once all this work is done: https://github.com/novuhq/novu/pull/4916
This PR needs to be done on top of it to inject the prefix.
